gpt4all-bindings/python/gpt4all/pyllmodel.py

```diff
@@ -237,6 +239,82 @@ def generate(self,
         print()
         return stream_processor.output
 
+    def generator(self, 
+                  prompt: str,
+                  logits_size: int = 0, 
+                  tokens_size: int = 0, 
+                  n_past: int = 0, 
+                  n_ctx: int = 1024, 
+                  n_predict: int = 128, 
+                  top_k: int = 40, 
+                  top_p: float = .9, 
+                  temp: float = .1, 
+                  n_batch: int = 8, 
+                  repeat_penalty: float = 1.2, 
+                  repeat_last_n: int = 10, 
+                  context_erase: float = .5) -> str:
+
+        # Symbol to terminate from generator
+        TERMINATING_SYMBOL = "#TERMINATE#"
+        
+        output_queue = queue.Queue()
+
+        prompt = prompt.encode('utf-8')
+        prompt = ctypes.c_char_p(prompt)
+
+        context = LLModelPromptContext(
+            logits_size=logits_size, 
+            tokens_size=tokens_size, 
+            n_past=n_past, 
+            n_ctx=n_ctx, 
+            n_predict=n_predict, 
+            top_k=top_k, 
+            top_p=top_p, 
+            temp=temp, 
+            n_batch=n_batch, 
+            repeat_penalty=repeat_penalty, 
+            repeat_last_n=repeat_last_n, 
+            context_erase=context_erase
+        )
+
+        # Put response tokens into an output queue
+        def _generator_response_callback(token_id, response):
+            output_queue.put(response.decode('utf-8'))
```

Comment:
one problem with this is that we will not necessarily get valid UTF-8 after every token, for example an emoji output by llama or GPTJ will consist of multiple tokens that will only decode as valid utf8 after their decoded bytes are concatenated

we either want to yield raw bytes (if we want to keep a 1-1 relationship between token and generator item) or buffer bytes and only yield when they decode to valid utf8 (possibly fits some user expectations more but also violates the 1-item = 1-token expectation) or allow the user to choose from those - there is also an edge case with buffering - potentially generation could stop without ending on a valid utf8 decode

Reply:
Ah great call out! I'm not a huge fan of buffering given the issues you just mentioned.
This impacts the other methods too and if we switch to returning/yielding raw bytes I think this would confuse existing users for what seems like a somewhat rare scenario. I added `replace` so we can at least avoid any UnicodeDecode errors for now.

Reply:
#867 deals with that, though, no?